src/backend/distributed/transaction/distributed_transaction_management.c

```diff
@@ -0,0 +1,431 @@
+								   MaxBackends));
+
+	return size;
+}
+
+
+/*
+ *  InitializeDistributedTransactionManagementBackend is called per backend
+ *  and does the required initilizations.
+ */
+void
+InitializeDistributedTransactionManagementBackend(void)
+{
+	MyDistributedTransactionBackend =
+		&distributedTransactionShmemData->sessions[MyProc->pgprocno];
+
+	Assert(MyDistributedTransactionBackend);
+
+	SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+	MyDistributedTransactionBackend->databaseId = 0;
+	MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier = 0;
+	MyDistributedTransactionBackend->transactionId.transactionId = 0;
+	MyDistributedTransactionBackend->transactionId.timestamp = 0;
+
+	SpinLockRelease(&MyDistributedTransactionBackend->mutex);
+}
+
+
+/*
+ * UnSetDistributedTransactionId simply acquires the mutex and resets
+ * the backend's distributed transaction data in shared memory to the
+ * initial values.
+ */
+void
+UnSetDistributedTransactionId(void)
+{
+	if (MyDistributedTransactionBackend)
+	{
+		SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+		MyDistributedTransactionBackend->databaseId = 0;
+		MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier = 0;
+		MyDistributedTransactionBackend->transactionId.transactionId = 0;
+		MyDistributedTransactionBackend->transactionId.timestamp = 0;
+
+		SpinLockRelease(&MyDistributedTransactionBackend->mutex);
+	}
+}
+
+
+/*
+ * GetCurrentDistributedTransctionId reads the backend's distributed transaction id and
+ * returns a copy of it.
+ */
+DistributedTransactionId *
+GetCurrentDistributedTransctionId(void)
+{
+	DistributedTransactionId *currentDistributedTransactionId =
+		(DistributedTransactionId *) palloc(sizeof(DistributedTransactionId));
+
+	SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+	currentDistributedTransactionId->initiatorNodeIdentifier =
+		MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier;
+	currentDistributedTransactionId->transactionId =
+		MyDistributedTransactionBackend->transactionId.transactionId;
+	currentDistributedTransactionId->timestamp =
+		MyDistributedTransactionBackend->transactionId.timestamp;
+
+	SpinLockRelease(&MyDistributedTransactionBackend->mutex);
+
+	return currentDistributedTransactionId;
+}
+
+
+/*
+ * AssignDistributedTransactionId generates a new distributed transaction
+ * id and sets it for the current backend.
+ *
+ * This function should only be called on BeginCoordinatedTransaction(). Any other
+ * callers is very likely to break the distributed transction management.
+ */
+void
+AssignDistributedTransactionId(void)
+{
+	pg_atomic_uint64 *transactionIdSequence =
+		&distributedTransactionShmemData->nextTransactionId;
+
+	SpinLockAcquire(&MyDistributedTransactionBackend->mutex);
+
+	MyDistributedTransactionBackend->databaseId = MyDatabaseId;
+	MyDistributedTransactionBackend->transactionId.initiatorNodeIdentifier =
+		(uint64) GetLocalGroupId();
```

Comment:
Ah, the stupid error happened after I removed the extra functions, otherwise, I'm aware of those issues :/